pkg/kv/kvserver/replica_lifecycle_datadriven_test.go

```diff
@@ -257,36 +243,98 @@ func TestReplicaLifecycleDataDriven(t *testing.T) {
-
-					in := batcheval.SplitTriggerHelperInput{
-						LeftLease:      rs.lease,
-						GCThreshold:    &rs.gcThreshold,
-						GCHint:         &rs.gcHint,
-						ReplicaVersion: rs.version,
-					}
-					_, _, err := batcheval.TestingSplitTrigger(
-						ctx, rec, batch, enginepb.MVCCStats{}, split, in, hlc.Timestamp{},
-					)
-					require.NoError(t, err)
+				// Run the split trigger evaluation and capture the batch that's
+				// generated for replication. Stash it away. This represents the
+				// raft log entry that will be applied as part of split
+				// application.
+				batch := tc.storage.NewBatch()
+				rec := (&batcheval.MockEvalCtx{
+					ClusterSettings:        tc.st,
+					Desc:                   &desc,
+					Clock:                  tc.clock,
+					AbortSpan:              rs.abortspan,
+					LastReplicaGCTimestamp: rs.lastGCTimestamp,
+					RangeLeaseDuration:     tc.rangeLeaseDuration,
+				}).EvalContext()
+
+				in := batcheval.SplitTriggerHelperInput{
+					LeftLease:      rs.lease,
+					GCThreshold:    &rs.gcThreshold,
+					GCHint:         &rs.gcHint,
+					ReplicaVersion: rs.version,
+				}
+				_, _, err := batcheval.TestingSplitTrigger(
+					ctx, rec, batch, enginepb.MVCCStats{}, split, in, hlc.Timestamp{},
+				)
+				require.NoError(t, err)
+
+				tc.splits[rangeID] = &pendingSplit{
+					trigger: split,
+					batch:   batch,
+				}
+				tc.updatePostSplitRangeState(ctx, t, batch, split)
 
-					tc.updatePostSplitRangeState(t, ctx, batch, rangeID, split)
+				output, err := print.DecodeWriteBatch(batch.Repr())
+				require.NoError(t, err)
+				return strings.ReplaceAll(output, "\n\n", "\n")
+
+			case "apply-split":
+				rangeID := dd.ScanArg[roachpb.RangeID](t, d, "range-id")
+				ps, ok := tc.splits[rangeID]
+				require.True(t, ok, "pending split not found for range-id %d", rangeID)
+				split := ps.trigger
+
+				// Determine if the RHS is "destroyed" by checking replica
+				// state. this mirrors the logic in validateandpreparesplit:
+				// - if there's no replica for the rhs, it's considered
+				// destroyed.
+				// - if the replica has a higher ReplicaID than in the split
+				// trigger, the original was removed and a new one created (also
+				// destroyed).
+				rhsRangeState := tc.mustGetRangeState(t, split.RightDesc.RangeID)
+				rhsReplDesc := rhsRangeState.mustGetReplicaDescriptor(t, roachpb.NodeID(1))
+				destroyed := rhsRangeState.replica == nil ||
+					rhsRangeState.replica.ReplicaID > rhsReplDesc.ReplicaID
+
+				in := splitPreApplyInput{
+					destroyed:           destroyed,
+					rhsDesc:             split.RightDesc,
+					initClosedTimestamp: hlc.Timestamp{WallTime: 100}, // dummy timestamp
+				}
+
+				// Apply the stashed batch and run splitPreApply on top of it.
+				output := tc.mutate(t, func(batch storage.Batch) {
+					// First, apply the stashed batch from split trigger evaluation.
+					require.NoError(t, batch.ApplyBatchRepr(ps.batch.Repr(), false /* sync */))
```

Comment:
As mentioned in another comment, better store `batch.Repr()` in `ps` instead of computing it on the fly here.